File: smash/fcore/derived_type/mwd_atmos_data.f90
!%      (MWD) Module Wrapped and Differentiated.
!%
!%      Type
!%      ----
!%
!%      - Atmos_DataDT
!%           Atmospheric data used to force smash and derived quantities.
!%
!%          ======================== =======================================
!%          `Variables`              Description
!%          ======================== =======================================
!%          ``prcp``                 Precipitation field                         [mm]
!%          ``pet``                  Potential evapotranspiration field          [mm]
!%          ``snow``                 Snow field                                  [mm]
!%          ``temp``                 Temperature field                           [C]
!%          ``sparse_prcp``          Sparse precipitation field                  [mm]
!%          ``sparse_pet``           Sparse potential evapotranspiration field   [mm]
!%          ``sparse_snow``          Sparse snow field                           [mm]
!%          ``sparse_temp``          Sparse temperature field                    [C]
!%          ``mean_prcp``            Mean precipitation at gauge                 [mm]
!%          ``mean_pet``             Mean potential evapotranspiration at gauge  [mm]
!%          ``mean_snow``            Mean snow at gauge                          [mm]
!%          ``mean_temp``            Mean temperature at gauge                   [C]
!%          ======================== =======================================
!%
!%      Subroutine
!%      ----------
!%
!%      - Atmos_DataDT_initialise
!%      - Atmos_DataDT_copy

module mwd_atmos_data

    use md_constant !% only: sp
    use mwd_setup !% only: SetupDT
    use mwd_mesh !% only: MeshDT
    use mwd_sparse_matrix !% only: Sparse_MatrixDT, Sparse_MatrixDT_initialise_array

    implicit none

    type Atmos_DataDT

        real(sp), dimension(:, :, :), allocatable :: prcp
        real(sp), dimension(:, :, :), allocatable :: pet
        real(sp), dimension(:, :, :), allocatable :: snow
        real(sp), dimension(:, :, :), allocatable :: temp

        type(Sparse_MatrixDT), dimension(:), allocatable :: sparse_prcp
        type(Sparse_MatrixDT), dimension(:), allocatable :: sparse_pet
        type(Sparse_MatrixDT), dimension(:), allocatable :: sparse_snow
        type(Sparse_MatrixDT), dimension(:), allocatable :: sparse_temp

        real(sp), dimension(:, :), allocatable :: mean_prcp
        real(sp), dimension(:, :), allocatable :: mean_pet
        real(sp), dimension(:, :), allocatable :: mean_snow
        real(sp), dimension(:, :), allocatable :: mean_temp

    end type Atmos_DataDT

contains

    subroutine Atmos_DataDT_initialise(this, setup, mesh)

        implicit none

        type(Atmos_DataDT), intent(inout) :: this
        type(SetupDT), intent(in) :: setup
        type(MeshDT), intent(in) :: mesh

        if (setup%sparse_storage) then

            allocate (this%sparse_prcp(setup%ntime_step))
            call Sparse_MatrixDT_initialise_array(this%sparse_prcp, 0, .true., -99._sp)
            allocate (this%sparse_pet(setup%ntime_step))
            call Sparse_MatrixDT_initialise_array(this%sparse_pet, 0, .true., -99._sp)

            if (setup%snow_module_present) then
                allocate (this%sparse_snow(setup%ntime_step))
                call Sparse_MatrixDT_initialise_array(this%sparse_snow, 0, .true., -99._sp)
                allocate (this%sparse_temp(setup%ntime_step))
                call Sparse_MatrixDT_initialise_array(this%sparse_temp, 0, .true., -99._sp)
            end if

        else

            allocate (this%prcp(mesh%nrow, mesh%ncol, setup%ntime_step))
            this%prcp = -99._sp
            allocate (this%pet(mesh%nrow, mesh%ncol, setup%ntime_step))
            this%pet = -99._sp

            if (setup%snow_module_present) then
                allocate (this%snow(mesh%nrow, mesh%ncol, setup%ntime_step))
                this%snow = -99._sp
                allocate (this%temp(mesh%nrow, mesh%ncol, setup%ntime_step))
                this%temp = -99._sp
            end if

        end if

        allocate (this%mean_prcp(mesh%ng, setup%ntime_step))
        this%mean_prcp = -99._sp
        allocate (this%mean_pet(mesh%ng, setup%ntime_step))
        this%mean_pet = -99._sp

        if (setup%snow_module_present) then
            allocate (this%mean_snow(mesh%ng, setup%ntime_step))
            this%mean_snow = -99._sp
            allocate (this%mean_temp(mesh%ng, setup%ntime_step))
            this%mean_temp = -99._sp
        end if

    end subroutine Atmos_DataDT_initialise

    subroutine Atmos_DataDT_copy(this, this_copy)

        implicit none

        type(Atmos_DataDT), intent(in) :: this
        type(Atmos_DataDT), intent(out) :: this_copy

        this_copy = this

    end subroutine Atmos_DataDT_copy

end module mwd_atmos_data

File: Pytorch_playground/pycore/derived_type/mwd_atmos_data.py
import numpy as np
from pycore.derived_type.mwd_sparse_matrix import Sparse_MatrixDT

class Atmos_DataDT:
    """
    Class representing atmospheric data.

    Attributes:
        prcp (ndarray): Array representing precipitation data.
        pet (ndarray): Array representing potential evapotranspiration data.
        snow (ndarray): Array representing snow data.
        temp (ndarray): Array representing temperature data.
        sm (ndarray): Array representing soil moisture data.
        sparse_prcp (list): List of Sparse_MatrixDT objects representing sparse precipitation data.
        sparse_pet (list): List of Sparse_MatrixDT objects representing sparse potential evapotranspiration data.
        sparse_snow (list): List of Sparse_MatrixDT objects representing sparse snow data.
        sparse_temp (list): List of Sparse_MatrixDT objects representing sparse temperature data.
        sparse_sm (list): List of Sparse_MatrixDT objects representing sparse soil moisture data.
        mean_prcp (ndarray): Array representing mean precipitation data.
        mean_pet (ndarray): Array representing mean potential evapotranspiration data.
        mean_snow (ndarray): Array representing mean snow data.
        mean_temp (ndarray): Array representing mean temperature data.
        mean_sm (ndarray): Array representing mean soil moisture data.

    Methods:
        __init__(setup, mesh): Initializes an instance of Atmos_DataDT.
        copy(): Creates a copy of the Atmos_DataDT object.
    """

    def __init__(self, setup, mesh):
        """
        Initializes an instance of Atmos_DataDT.

        Args:
            setup (SetupDT): Setup object containing setup parameters.
            mesh (MeshDT): Mesh object containing mesh parameters.
        """
        if setup.sparse_storage:
            self.sparse_prcp = [Sparse_MatrixDT(0, True, -99.0) for _ in range(setup.ntime_step)]
            self.sparse_pet = [Sparse_MatrixDT(0, True, -99.0) for _ in range(setup.ntime_step)]
            if setup.snow_module_present:
                self.sparse_snow = [Sparse_MatrixDT(0, True, -99.0) for _ in range(setup.ntime_step)]
                self.sparse_temp = [Sparse_MatrixDT(0, True, -99.0) for _ in range(setup.ntime_step)]
            if setup.read_sm:
                self.sparse_sm = [Sparse_MatrixDT(0, True, -99.0) for _ in range(setup.ntime_step)]
        else:
            self.prcp = np.full((mesh.nrow, mesh.ncol, setup.ntime_step), -99.0)
            self.pet = np.full((mesh.nrow, mesh.ncol, setup.ntime_step), -99.0)
            if setup.snow_module_present:
                self.snow = np.full((mesh.nrow, mesh.ncol, setup.ntime_step), -99.0)
                self.temp = np.full((mesh.nrow, mesh.ncol, setup.ntime_step), -99.0)
            if setup.read_sm:
                self.sm = np.full((mesh.nrow, mesh.ncol, setup.ntime_step), -99.0)

        self.mean_prcp = np.full((mesh.ng, setup.ntime_step), -99.0)
        self.mean_pet = np.full((mesh.ng, setup.ntime_step), -99.0)
        if setup.snow_module_present:
            self.mean_snow = np.full((mesh.ng, setup.ntime_step), -99.0)
            self.mean_temp = np.full((mesh.ng, setup.ntime_step), -99.0)
        if setup.read_sm:
            self.mean_sm = np.full((mesh.ng, setup.ntime_step), -99.0)

    def copy(self):
        """
        Creates a copy of the Atmos_DataDT object.

        Returns:
            Atmos_DataDT: Copy of the Atmos_DataDT object.
        """
        return Atmos_DataDT(self)